### DISCUSSION TITLE
A single "meta" export for static frontmatter

### DISCUSSION THREAD

**Discussion by bholmesdev** [withastro/roadmap#135] [Proposal]
I've noticed a few RFCs calling for "special" exports that Astro can pick up on. To share a few example:
- We need a way to flag pages for SSR (#133). Maybe we could use an export like `export const staticRender = true`?
- We might want to defer certain pages for deprioritized builds (#127). Maybe we could use an export like `export const defer = X`?
- We may want control over cache policies, request headers, and other handler options when deploying to SSR. Maybe a magical object like `export const headers = {...}` would do the trick?

I could see this ballooning to a _plethora_ of magic exports that Astro expects. 

## The problem with export excitement

There's a couple pain points here:
- **each new export further pollutes the namespace of your templates.** If we eventually support named exports from `.astro` files similar to [Svelte's `context="module"`](https://svelte.dev/tutorial/module-exports) (which could be super useful on `fetchContent` #7 👀 ), it'll get harder and harder to introduce "reserved" exports overtime.
- **it's labor-intensive to add types.** We'll ideally export type definitions for each first-class export you use. This causes more complex import statements overtime:

```ts
import type { Headers, RenderMode, Defer } from 'astro'

export const renderMode: RenderMode = 'ssr'
export const defer: Defer = 0
export const headers: Headers = {...}
```

This is _fine..._ but to be completely honest, I'd struggle to remember all these import / export names. Ideally, I could import a single type that shows me _all_ available flags.

## Enter the `meta` object

What we want is a **single, statically analyzable object** for every page on your site. [Using Remix for inspiration](https://remix.run/docs/en/v1/api/conventions#meta), I propose we add a new `meta` export.

Let's rewrite that export bonanza from earlier into a single `meta` object:

```ts
import type { Meta } from 'astro'

export const meta: Meta = {
  renderMode: 'ssr',
  defer: 0,
  headers: {...},
}
```

There's a few important pieces to note:
1. Instead of maintaining a suite of types, we can maintain a single `Meta` type.
2. As the `Meta` type updates overtime, your editor's intellisense will update as well. This is great for discoverability without reading the release notes
3. Your namespace isn't polluted anymore! If Astro ever introduces a new, statically analyzable piece of data, we can place it on the `meta` object
4. It's named after a very successful VR compan- okay maybe that's not a good reason 😉 

## A future for the integrations layer?

I also think the `meta` object would help with custom integrations at the config. I'll wait for @FredKSchott's full proposal to discuss details, but I'll mention a rough example here: opt-in to injecting Tailwind styles on a page-by-page basis.

```astro
---
// unstyled-page.astro
import type { Meta } from 'astro'
import type { TailwindMeta } from '@astrojs/tailwind'

export const meta: Meta & TailwindMeta = {
  injectedStyles: {
    tailwind: false,
  }
}
---
...
```

```ts
// @astrojs/tailwind/config.ts
hooks: {
  'astro:build:start': ({ configByPage }) => {
    configByPage(({ meta, inject }) => {
      if (meta.injectedStyles.tailwind === false) return
      inject(/**tailwind stuff*/)
    })
  }
}
```

☝️ This API is _super_ unpolished, sure. But it conveys a future need among integration authors: opting in to certain options page-by-page using metadata.

## In summation

A meta object is:
- simple for the developer to understand
- simple for Astro to maintain / update overtime
- simple for our bundler to pluck out at compile time

I recommend this approach for flags going forward!

**Comment by FredKSchott** [comment]
I really like the concept! It would be good to have a really clear use-case in mind for this to discuss. Maybe SSR with a `revalidate` key to control caching? 

One thing to keep in mind is that exports in the front matter script are hoisted out of the render and into a top-level context, which means that we document them today as "hey, you can't access anything outside of `getStaticPaths()` other than imports. any variable references will fail due to hoisting." 

Is this an acceptable limitation for what you had in mind? To use `revalidate` or `tailwind` as an example again, it would mean that you could never control the value based on any dynamic value and would need to use a static value for all pages.

**Reply by bholmesdev** [reply]
I think the `staticRender` flag from #133 will be the most immediate use case!

And yes, that's definitely the limitation I had in mind. Meta values should be pre-compile-able so use cases like config and `fetchContent` could scan values without rendering anything. I _did_ notice that [NextJS' revalidate flag](https://nextjs.org/docs/api-reference/data-fetching/get-static-props#revalidate) is compute-able though via `getStaticProps`. Maybe `meta` could be a synchronous or asynchronous function if the user prefers? I'd expect most of our documented examples to use a plain, frontmatter-style object though.
